pyannote/audio/embedding/approaches/contrastive_loss.py

```diff
@@ -28,85 +28,68 @@
+from .base import RepresentationLearning
 
 
-class ContrastiveLoss(EmbeddingApproach):
-    """Train embeddings using triplet loss
+class ContrastiveLoss(RepresentationLearning):
+    """Contrasitve loss
+
+    TODO explain
 
     Parameters
     ----------
     duration : float, optional
         Chunks duration, in seconds. Defaults to 1.
+    per_turn : int, optional
+        Number of chunks per speech turn. Defaults to 1.
+        If per_turn is greater than one, embeddings of the same speech turn
+        are averaged before comparison. The intuition is that it might help
+        learn embeddings meant to be averaged/summed.
+    per_label : `int`, optional
+        Number of sequences per speaker in each batch. Defaults to 1.
+    per_fold : `int`, optional
+        Number of different speakers per batch. Defaults to 32.
+    per_epoch : `float`, optional
+        Force total audio duration per epoch, in days.
+        Defaults to total duration of protocol subset.
+    label_min_duration : `float`, optional
+        Remove speakers with less than that many seconds of speech.
+        Defaults to 0 (i.e. keep them all).
     metric : {'euclidean', 'cosine', 'angular'}, optional
         Defaults to 'cosine'.
     margin: float, optional
         Margin multiplicative factor. Defaults to 0.2.
-    size_average : `bool`, optional
-        Divide total loss by batch size
-    per_label : int, optional
-        Number of sequences per speaker in each batch. Defaults to 3.
-    label_min_duration : float, optional
-        Remove speakers with less than `label_min_duration` seconds of speech.
-        Defaults to 0 (i.e. keep them all).
-    per_fold : int, optional
-        If provided, sample triplets from groups of `per_fold` speakers at a
-        time. Defaults to sample triplets from the whole speaker set.
-    per_epoch : float, optional
-        Force total audio duration per epoch, in days.
-        Defaults to total duration of protocol subset.
+
+    Reference
+    ---------
+    TODO
     """
 
-    def __init__(self, duration=1.0,
-                       metric='cosine',
-                       margin=0.2,
-                       size_average=True,
-                       per_label=3,
-                       per_fold=None,
+    def __init__(self, duration: float = 1.0,
+                       per_turn: int = 1,
+                       per_label: int = 1,
+                       per_fold: int = 32,
                        per_epoch: float = None,
-                       label_min_duration=0.):
-
-        super().__init__()
+                       label_min_duration: float = 0.,
+                       # FIXME create a Literal type for metric
+                       # FIXME maybe in pyannote.core.utils.distance
+                       metric: str = 'cosine',
+                       # FIXME homogeneize the meaning of margin parameter
+                       # FIXME it has a different meaning in ArcFace, right?
```

Comment:
yes, the contrastive loss margin specifies how far away the negative pairs should be, whereas the arcface margin specifies how far away we can tolerate positive pairs to be.
About homogenizing the margins, I don't know if it's a good idea because the definition of the margin is inherent to the definition of the loss, like in the example of arcface and contrastive. I think it would be easier when trying to reproduce experiments from papers. For instance, a 0.2 margin in a paper about contrastive loss would mean `ContrastiveLoss(margin=0.2)` in pyannote, otherwise we'd have to calculate which pyannote margin is equivalent to the other one.